File: Algorithm/nowcoder/maxLongSubArray.js
/**
 *
 * @param arr int整型一维数组 the array
 * @return int整型
 */
function maxLength(arr) {
  // write code here
  function getIndexByItem(iterator, item) {
    let index = 0;
    while (iterator.next().value) {
      index++;
      if (iterator.next().value === item) break;
    }
    return index;
  }
  let L = 0, R = 0, Len = arr.length;
  let maxValue = Number.MIN_VALUE;
  let unreSet = new Set();
  for (let i = 0;i < Len;i++) {
    if (!unreSet.has(arr[i])) {
      unreSet.add(arr[i])
      R++;
    } else {
      maxValue = maxValue > R - L ? maxValue : R - L
      //             L = R - 1;
      const step = getIndexByItem(unreSet.keys(), arr[i])
      L = L + step;
      unreSet.clear();
      unreSet.add(arr[i])
    }
    maxValue = maxValue > R - L ? maxValue : R - L
  }
  return maxValue;
}

console.log(maxLength([2, 2, 3, 4, 8, 99, 3]))

File: Algorithm/nowcoder/maxLongSubArray.go
package main

/**
 *
 * @param arr int整型一维数组 the array
 * @return int整型
*/
func maxLength( arr []int ) int {
    // write code here
    res := 0
    s := make([]int, 100010)
    j := 0
    for i := 0; i < len(arr); i ++ {
        s[arr[i]] ++
        for j <= i && s[arr[i]] > 1 {
            s[arr[j]] --
            j ++
        }
        if i - j + 1 > res {
            res = i - j + 1;
        }
    }
    return res
}